File: convert-sorted-array-to-binary-search-tree/Solution108.java
// https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/

class Solution108 {

    // T: O(n), S: O(logn)
    public TreeNode sortedArrayToBST(int[] nums) {
        return helper(nums, 0, nums.length - 1);
    }

    private TreeNode helper(int[] nums, int left, int right) {
        if (left > right) {
            return null;
        }
        int mid = left + (right - left) / 2;
        var node = new TreeNode(nums[mid]);
        node.left = helper(nums, left, mid - 1);
        node.right = helper(nums, mid + 1, right);
        return node;
    }

    public static void main(String[] args) {
        var sol = new Solution108();
        var nums1 = new int[] { -10, -3, 0, 5, 9 };
        var nums2 = new int[] { 1, 3 };
        var root1 = sol.sortedArrayToBST(nums1);
        System.out.println(root1); // [0,-10,5,null,-3,null,9]
        var root2 = sol.sortedArrayToBST(nums2);
        System.out.println(root2); // [1,null,3]
    }

}

File: convert-sorted-array-to-binary-search-tree/solution108.go
// https://leetcode.com/problems/convert-sorted-array-to-binary-search-tree/

package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// T: O(n), S: O(logn)
func sortedArrayToBST(nums []int) *TreeNode {
	return helper(nums, 0, len(nums)-1)
}

func helper(nums []int, left, right int) *TreeNode {
	if left > right {
		return nil
	}
	mid := left + (right-left)/2
	node := &TreeNode{Val: nums[mid]}
	node.Left = helper(nums, left, mid-1)
	node.Right = helper(nums, mid+1, right)
	return node
}

func main() {
	nums1 := []int{-10, -3, 0, 5, 9}
	nums2 := []int{1, 3}
	root1 := sortedArrayToBST(nums1)
	fmt.Println(root1) // [0,-10,5,null,-3,null,9]
	root2 := sortedArrayToBST(nums2)
	fmt.Println(root2) // [1,null,3]
}
